### ISSUE TITLE
Automatically deriving implementations

### ISSUE THREAD

**Issue by vandenheuvel** [brendanzab/approx#43]
Is there any outlook on automatic derivations of the traits in this crate?

I tried to write a derive crate, but got stuck. Deriving for example `AbsDiffEq` for structs could be done by calling `AbsDiffEq::abs_diff_eq` on all fields, in combination with an `impl<T: Eq> AbsDiffEq for T`. This last trait implementation collides however with the existing derivation implementations such as `impl<'a, T: AbsDiffEq> AbsDiffEq for &'a T`. I don't know how to solve this problem without removing these trait derivations, or until [this work](https://github.com/rust-lang/rfcs/pull/1210) is finished.

**Comment by tylerreisinger** [comment]
Is the last blanket impl necessary? You could always just create a macro to create the implementation for foreign types if needed

**Comment by tylerreisinger** [comment]
Here's a pretty quick and dirty example of the sort of thing needed: https://gist.github.com/tylerreisinger/886e56379beffabddaff7bad6a232f06. It can only handles structs and won't insert type bounds on generics, but otherwise works.

**Comment by tylerreisinger** [comment]
Also, there is no clean way to deal with both non-primitive Epsilon types and a mix of f32/f64 types without TryFrom, which is currently unstable. I used `as` in my example, as I'm guessing non-primitive Epsilons aren't *that* common.

**Comment by aldanor** [comment]
A nice thing to have would be a derive-macro, so you could do

```rust
#[derive(AbsDiffEq)]
struct MyStruct { ... }
```

**Comment by joshburkart** [comment]
This would be really great to have! Approximately comparing nested `struct`s/`enum`s is a pretty common use case when writing tests for numerical computing applications. @tylerreisinger Do you think the implementation you made is ready for a PR in some form? Or too hacky right now?

**Comment by tylerreisinger** [comment]
@joshburkart It's not ready for a PR yet, but I can work on it a bit this weekend and see if I can get something working. TryFrom being stabilized helps with an issue implementing it.

**Comment by brendanzab** [comment]
Yeah, having this would be really nice! I know it'll be a bit annoying with the epsilon though, and haven't had much time at the moment to look into it.

**Comment by tylerreisinger** [comment]
I've created a repository at https://github.com/tylerreisinger/approx_derive to implement it. When it is production ready, I can transfer ownership or merge it into the approx repo with a cargo workspace.

**Comment by ThatsGobbles** [comment]
I'd love to have this as well!

**Comment by Sufflope** [comment]
Big woops, didn't see https://github.com/jonaspleyer/approx-derive created between my first usage of `approx` 2 years ago, and my latest one month ago when I tought "come on it needs some derives" :see_no_evil: 

Some pretty big differences in design choices so I'll see how to best contribute to it, but at least linking it here for visibility to at least avoid having a third person going full xkcd 927 :sweat_smile:

**Comment by jonaspleyer** [comment]
Hey everyone, I have just recently updated my `approx-derive` crate with some new features. Is there an effort or desire to merge the two crates? As far as I can tell, the maintainer of `approx` has been inactive for a longer period of time.

**Comment by jonaspleyer** [comment]
To summarize the functionalities of my crate. It solves some of the problems already encountered by users:
- Use equality in fields where desired
- Custom `epsilon` values for individual fields
- Custom `max_relative` value
- Casting of values (such as `f32` to `f64` or vice versa)
- Skip individual fields
- Generics are supported (see [tests](https://github.com/jonaspleyer/approx-derive/blob/main/tests/derive_abs_diff.rs#L120))
- Custom mapping of values before comparing them (eg. map circle area to radius `|v| v.sqrt() / f32::consts::PI / 2.0`).

**Comment by jonaspleyer** [comment]
This has been implemented in the most recent version `0.6.0-rc1` please check that this version satisfies your needs.
